### PR TITLE
Introduce Mvc analyzers

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: build/dependencies.props [hunk 1]
```diff
@@ -42,6 +42,7 @@
     <MicrosoftAspNetCoreWebUtilitiesPackageVersion>2.1.0-preview1-28055</MicrosoftAspNetCoreWebUtilitiesPackageVersion>
     <MicrosoftAspNetWebApiClientPackageVersion>5.2.4-preview1</MicrosoftAspNetWebApiClientPackageVersion>
     <MicrosoftCodeAnalysisCSharpPackageVersion>2.3.1</MicrosoftCodeAnalysisCSharpPackageVersion>
+    <MicrosoftCodeAnalysisCSharpWorkspacesPackageVersion>2.6.1</MicrosoftCodeAnalysisCSharpWorkspacesPackageVersion>
```
Comment (rynowak) [inline build/dependencies.props]: This isn't right. This should be the same version as all of the other Roslyn packages.

I assume this is for tests right?

File: src/Microsoft.AspNetCore.Mvc.Analyzers/ActionsMustNotBeAsyncVoidAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.AspNetCore.Mvc.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class ActionsMustNotBeAsyncVoidAnalyzer : ControllerAnalyzerBase
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Analyzers/ActionsMustNotBeAsyncVoidAnalyzer.cs]: I think this one is a sure winner

File: src/Microsoft.AspNetCore.Mvc.Analyzers/ApiControllerAnalyzerContext.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        public INamedTypeSymbol ObjectResult => GetType(TypeNames.ObjectResult, ref _objectResult);
+
+        private INamedTypeSymbol _iActionResult;
+        public INamedTypeSymbol IActionResult => GetType(TypeNames.IActionResult, ref _iActionResult);
+
+        public INamedTypeSymbol _modelState;
+        public INamedTypeSymbol ModelStateDictionary => GetType(TypeNames.ModelStateDictionary, ref _modelState);
+
+        public INamedTypeSymbol _nonActionAttribute;
+        public INamedTypeSymbol NonActionAttribute => GetType(TypeNames.NonActionAttribute, ref _nonActionAttribute);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Analyzers/ApiControllerAnalyzerContext.cs]: BTW I'm not really certain that caching these has value. I think they build a big table of names -> symbols

File: src/Microsoft.AspNetCore.Mvc.Analyzers/ApiControllerFacts.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Mvc.Analyzers
+{
+    public static class ApiControllerFacts
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Analyzers/ApiControllerFacts.cs]: LEL?

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Analyzers/ApiControllerFacts.cs]: If we're not actually testing anything, let's not add this project. We have plenty of projects

File: src/Microsoft.AspNetCore.Mvc.Analyzers/CodeAnalysisExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Mvc.Analyzers
+{
+    internal static class CodeAnalysisExtensions
+    {
+        public static bool HasAttribute(this ITypeSymbol typeSymbol, ITypeSymbol attribute, bool inherit)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Analyzers/CodeAnalysisExtensions.cs]: I'm not a fan of extensions in general but these seem like real gaps in Roslyn's api - unless this code is wrong

File: test/Microsoft.AspNetCore.Mvc.Analyzers.Test/ActionsMustNotBeAsyncVoidFacts.cs [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.Analyzers.Infrastructure;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.Analyzers
+{
+    public class ActionsMustNotBeAsyncVoidFacts : AnalyzerTestBase
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Analyzers.Test/ActionsMustNotBeAsyncVoidFacts.cs]: Ah I see, you did make some tests

File: src/Microsoft.AspNetCore.Mvc.Analyzers/ApiControllerAnalyzerContext.cs (hunk 3 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Analyzers/ApiControllerAnalyzerContext.cs]: I hadn't measured it, but when I dug through their code, it didn't look like they cached the results of their lookup - https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs#L550-L585

They also cache this as state so it's evaluated once per analyzer - https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/UseCollectionInitializer/AbstractUseCollectionInitializerDiagnosticAnalyzer.cs#L46-L55

File: src/Microsoft.AspNetCore.Mvc.Analyzers/ApiControllerFacts.cs (hunk 4 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Analyzers/ApiControllerFacts.cs]: Oops

File: test/Microsoft.AspNetCore.Mvc.Analyzers.Test/Infrastructure/AnalyzerTestBase.cs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            var projectId = ProjectId.CreateNewId(debugName: "TestProject");
+            var newFileName = "Test.cs";
+            var documentId = DocumentId.CreateNewId(projectId, debugName: newFileName);
+            var metadataReferences = DependencyContext.Default.CompileLibraries
+                .SelectMany(c => c.ResolveReferencePaths())
+                .Select(path => MetadataReference.CreateFromFile(path))
+                .Cast<MetadataReference>()
+                .ToList();
+
+            Workspace = Workspace ?? new AdhocWorkspace();
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Analyzers.Test/Infrastructure/AnalyzerTestBase.cs]: OH! Btw all creation of an `AdhocWorkspace` needs to be protected by a global (process-level) lock. There is a bug in MEF on desktop framework that we hit recently in razor.

Comment (pranavkm) [reply test/Microsoft.AspNetCore.Mvc.Analyzers.Test/Infrastructure/AnalyzerTestBase.cs]: Ahh, I saw these errors randomly and I thought I had a test bug (the tests generated by the analyzer template have a lot of Task.Result). Ok, I'll see what you guys did in Razor

File: version.props [hunk 8]
```diff
@@ -2,9 +2,15 @@
   <PropertyGroup>
     <VersionPrefix>2.1.0</VersionPrefix>
     <VersionSuffix>preview1</VersionSuffix>
+
+    <ExperimentalVersionPrefix>0.1.0</ExperimentalVersionPrefix>
```
Comment (pranavkm) [inline version.props]: FYI @natemcmaster

File: src/Microsoft.AspNetCore.Mvc.Analyzers/Microsoft.AspNetCore.Mvc.Analyzers.csproj [hunk 9]
```diff
@@ -0,0 +1,24 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <Description>CSharp Analyzers for ASP.NET Core MVC.</Description>
+    <PackageTags>aspnetcore;aspnetcoremvc</PackageTags>
+
+    <VerifyVersion>false</VerifyVersion>
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Analyzers/Microsoft.AspNetCore.Mvc.Analyzers.csproj]: FYI @natemcmaster

**Pull request merged**